File: ts/src/suggestion_engine/rules/generation.ts
import { SEVERITY_ERROR, SEVERITY_WARNING } from "../../reviewer/types.ts";
import type { Verdict } from "../../reviewer/types.ts";
import type { SuggestionRule } from "../protocol.ts";
import type { Suggestion } from "../types.ts";
import type { EvalResult } from "../../types.ts";

export class LowSuccessRateRule implements SuggestionRule {
  readonly categories = new Set(["generation"]);
  private readonly threshold: number;

  constructor(threshold = 0.8) {
    this.threshold = threshold;
  }

  evaluate(result: EvalResult): Suggestion[] {
    if (result.metrics.pAt5 >= this.threshold) return [];
    return [
      {
        severity: SEVERITY_WARNING,
        message: `Generation success rate is ${Math.round(result.metrics.pAt5 * 100)}%, below threshold ${Math.round(this.threshold * 100)}%`,
        ruleName: "low_success_rate",
        metric: "success_rate",
        category: result.category,
        fixHint: "Check adapter error handling and API connectivity.",
      },
    ];
  }
}

export class HallucinationSignalRule implements SuggestionRule {
  readonly categories = new Set(["generation"]);
  private readonly scoreThreshold: number;

  constructor(scoreThreshold = 0.5) {
    this.scoreThreshold = scoreThreshold;
  }

  evaluate(
    result: EvalResult,
    verdicts?: Verdict[],
  ): Suggestion[] {
    if (!verdicts || verdicts.length === 0) return [];

    const flagged = verdicts.filter(
      (v) => !v.passed && v.score < this.scoreThreshold,
    );
    if (flagged.length === 0) return [];

    return [
      {
        severity: SEVERITY_ERROR,
        message: `${flagged.length} verdict(s) scored below ${this.scoreThreshold.toFixed(1)}, possible hallucination or fabricated content`,
        ruleName: "hallucination_signal",
        metric: "judge_score",
        category: result.category,
        fixHint:
          "Add grounding context to generation prompts; consider retrieval-augmented generation or stricter guardrails.",
      },
    ];
  }
}

export class VerdictFailureRule implements SuggestionRule {
  evaluate(
    result: EvalResult,
    verdicts?: Verdict[],
  ): Suggestion[] {
    if (!verdicts || verdicts.length === 0) return [];

    const suggestions: Suggestion[] = [];
    for (const verdict of verdicts) {
      for (const check of verdict.checks) {
        if (check.passed) continue;
        suggestions.push({
          severity: check.severity,
          message: `Check '${check.name}' failed: ${check.reasoning}`,
          ruleName: "verdict_failure",
          category: result.category,
        });
      }
    }
    return suggestions;
  }
}

File: src/synapt_eval/suggestion_engine/rules/generation.py
"""Generation suggestion rules."""

from __future__ import annotations

from typing import Any

from synapt_eval.reviewer.types import SEVERITY_ERROR, SEVERITY_WARNING, Verdict
from synapt_eval.suggestion_engine.types import Suggestion
from synapt_eval.types import EvalResult


class LowSuccessRateRule:
    """Flags when generation success rate drops below threshold."""

    categories = {"generation"}

    def __init__(self, threshold: float = 0.8) -> None:
        self._threshold = threshold

    def evaluate(
        self,
        result: EvalResult,
        verdicts: list[Verdict] | None = None,
        context: dict[str, Any] | None = None,
    ) -> list[Suggestion]:
        if result.metrics.p_at_5 >= self._threshold:
            return []
        return [
            Suggestion(
                severity=SEVERITY_WARNING,
                message=(
                    f"Generation success rate is {result.metrics.p_at_5:.0%}, "
                    f"below threshold {self._threshold:.0%}"
                ),
                rule_name="low_success_rate",
                metric="success_rate",
                category=result.category,
                fix_hint="Check adapter error handling and API connectivity.",
            )
        ]


class HallucinationSignalRule:
    """Flags when judge verdicts indicate hallucinated content.

    Scans verdicts for judge checks with low scores, which may
    indicate the generation fabricated content not grounded in context.
    """

    categories = {"generation"}

    def __init__(self, score_threshold: float = 0.5) -> None:
        self._score_threshold = score_threshold

    def evaluate(
        self,
        result: EvalResult,
        verdicts: list[Verdict] | None = None,
        context: dict[str, Any] | None = None,
    ) -> list[Suggestion]:
        if not verdicts:
            return []

        flagged = [
            v
            for v in verdicts
            if not v.passed and v.score is not None and v.score < self._score_threshold
        ]

        if not flagged:
            return []

        return [
            Suggestion(
                severity=SEVERITY_ERROR,
                message=(
                    f"{len(flagged)} verdict(s) scored below {self._score_threshold:.1f}, "
                    f"possible hallucination or fabricated content"
                ),
                rule_name="hallucination_signal",
                metric="judge_score",
                category=result.category,
                fix_hint=(
                    "Add grounding context to generation prompts; "
                    "consider retrieval-augmented generation or stricter guardrails."
                ),
            )
        ]


class VerdictFailureRule:
    """Creates suggestions from failed verdict checks.

    Generic rule that scans all verdict check results for failures
    and generates per-failure suggestions. Covers domain-specific
    checks (temporal drift, persona inconsistency, etc.) when the
    customer's reviewer chain includes those predicates.
    """

    def evaluate(
        self,
        result: EvalResult,
        verdicts: list[Verdict] | None = None,
        context: dict[str, Any] | None = None,
    ) -> list[Suggestion]:
        if not verdicts:
            return []

        suggestions: list[Suggestion] = []
        for verdict in verdicts:
            for check in verdict.checks:
                if check.passed:
                    continue
                suggestions.append(
                    Suggestion(
                        severity=check.severity,
                        message=f"Check '{check.name}' failed: {check.reasoning}",
                        rule_name="verdict_failure",
                        category=result.category,
                    )
                )

        return suggestions
